### PR TITLE
session persister bugs

### PR DESCRIPTION
**Pull request by jmazzitelli**
(_this is a WIP - I'm still not sure if these are legit bugs, but there is smoke here, so potentially we have a fire_)

fixes: https://github.com/kiali/kiali/issues/8990

# CookieSessionPersistor Bug Analysis Confirmation

## Overview

Analysis of bugs in `handlers/authentication/session_persistor.go` related to cookie naming and handling that can cause session corruption in multi-session and chunked session scenarios.

**File Analyzed:** `kiali/handlers/authentication/session_persistor.go`

---

### Problem

The filter in `ReadAllSessions` (line 336) incorrectly matches chunk cookies and tries to process them as full sessions:

```go
if (strings.HasPrefix(cookie.Name, SessionCookieName) || strings.HasPrefix(cookie.Name, NumberOfChunksCookieName)) && !strings.HasPrefix(cookie.Name, NonceCookieName) {
```

**Cookie name constants:**
- `SessionCookieName = "kiali-token"`
- `NumberOfChunksCookieName = "kiali-token-chunks"`
- `NonceCookieName = "kiali-token-nonce"`

**For a chunked session with empty key, the cookies are:**
- `kiali-token` (main session cookie)
- `kiali-token-1` (chunk 1)
- `kiali-token-chunks` = "2" (chunks count)

**The filter `strings.HasPrefix(cookie.Name, "kiali-token")` matches:**

| Cookie Name | Matches Filter | Should Be Processed |
|-------------|----------------|---------------------|
| `kiali-token` | ✓ | ✓ (correct - main session) |
| `kiali-token-1` | ✓ | ✗ **BUG** - chunk cookie |
| `kiali-token-chunks` | ✓ | ✗ **BUG** - count cookie |
| `kiali-token-cluster1-1` | ✓ | ✗ **BUG** - keyed chunk cookie |

### Test Evidence

```
=== RUN   TestBug1_ReadAllSessions_DropsChunkCookies
Reading session cookie: kiali-token-1
Error reading session cookie kiali-token-1: failed to read session cookie chunk number 1: http: named cookie not present
Dropping cookie: kiali-token-1
Reading session cookie: kiali-token-chunks
Error reading session cookie kiali-token-chunks: failed to read session cookie chunk number 1: http: named cookie not present
Dropping cookie: kiali-token-chunks

BUG CONFIRMED: ReadAllSessions dropped cookies: [kiali-token-1 kiali-token-chunks]
These cookies were incorrectly processed as full sessions and dropped when decryption failed
```

### Impact

- **Session corruption**: Chunk cookies are dropped when `readKialiCookie` fails to decrypt them
- **Subsequent requests fail**: Main session cookie references chunks that no longer exist
- **Multi-cluster scenarios**: More severe impact when multiple keyed sessions exist

### Suggested Fix

Add a helper function to identify main session cookies vs chunk cookies:

```go
// isMainSessionCookie returns true if the cookie name is a main session cookie
// (not a chunk cookie, chunks count cookie, or nonce cookie).
func isMainSessionCookie(cookieName string) bool {
    // Filter out nonce cookies
    if strings.Contains(cookieName, "nonce") {
        return false
    }

    // Check if it's a chunks count cookie
    if strings.HasPrefix(cookieName, NumberOfChunksCookieName) {
        return false
    }

    // Must start with session cookie name
    if !strings.HasPrefix(cookieName, SessionCookieName) {
        return false
    }

    // If it's exactly the session cookie name (no key), it's a main session
    if cookieName == SessionCookieName {
        return true
    }

    // Get the suffix after "kiali-token-"
    suffix := strings.TrimPrefix(cookieName, SessionCookieName+"-")

    // Check if the suffix ends with a numeric chunk number
    parts := strings.Split(suffix, "-")
    lastPart := parts[len(parts)-1]

    // If lastPart is purely numeric, this is a chunk cookie
    if _, err := strconv.Atoi(lastPart); err == nil {
        return false
    }

    return true
}
```

---

### Problem

The chunks count cookie is always named `kiali-token-chunks` regardless of the session key:

| Function | Code | Cookie Name Used |
|----------|------|------------------|
| `CreateSession` (line 192) | `Name: NumberOfChunksCookieName` | `"kiali-token-chunks"` |
| `readKialiCookie` (line 224) | `r.Cookie(NumberOfChunksCookieName)` | `"kiali-token-chunks"` |
| `TerminateSession` (line 387) | `sessionCookieName(NumberOfChunksCookieName, key)` | `"kiali-token-chunks-{key}"` |

### Test Evidence

**Test 2a - CreateSession doesn't key the chunks cookie:**
```
Expected chunks cookie name: kiali-token-chunks-cluster1
Actual chunks cookie name: kiali-token-chunks
BUG CONFIRMED: Chunks cookie is not keyed per session.
```

**Test 2b - TerminateSession can't find the chunks cookie:**
```
Created cookies: [kiali-token-cluster1 kiali-token-cluster1-1 kiali-token-chunks]
Dropped cookies: [kiali-token-cluster1 kiali-token-cluster1-1]
BUG CONFIRMED: TerminateSession did not drop the chunks cookie "kiali-token-chunks"
because it was looking for "kiali-token-chunks-cluster1" which doesn't exist
```

**Test 2c - Multiple sessions collide:**
```
Cluster1 chunks cookie value: 2
Cluster2 chunks cookie value: 3
BUG CONFIRMED: Both sessions use the same chunks cookie name "kiali-token-chunks".
Cluster1 has 2 chunks, cluster2 has 3 chunks.
The browser will only keep the latest value, causing session corruption.
```

### Impact

1. **Cleanup failure**: `TerminateSession` looks for a keyed cookie that was never created
2. **Session collision**: Multiple keyed sessions share the same `kiali-token-chunks` cookie
3. **Incorrect chunk reading**: Reading one session may use another session's chunk count

### Suggested Fix

Key the chunks count cookie per session:

**In `CreateSession`:**
```go
chunksCookie := http.Cookie{
    Name:     sessionCookieName(NumberOfChunksCookieName, s.Key),
    // ...
}
```

**In `readKialiCookie`:**
```go
sessionKey := extractKeyFromSessionCookieName(cookieName)
chunksCookieName := sessionCookieName(NumberOfChunksCookieName, sessionKey)
numChunksCookie, chunksCookieErr := r.Cookie(chunksCookieName)
```

Add helper function:
```go
func extractKeyFromSessionCookieName(cookieName string) string {
    if cookieName == SessionCookieName {
        return ""
    }
    return strings.TrimPrefix(cookieName, SessionCookieName+"-")
}
```

---

### Problem

The comments at multiple locations reference a "-aes" suffix that is no longer used in the code.

**Comment at lines 208-210:**
```go
// This CookieSessionPersistor only deals with sessions using cookies holding encrypted data.
// Thus, presence for a cookie with the "-aes" suffix is checked and it's assumed no active session
// if such cookie is not found in the request.
```

**Comment at lines 296-298:**
```go
// This CookieSessionPersistor only deals with sessions using cookies holding encrypted data.
// Thus, presence for a cookie with the "-aes" suffix is checked and it's assumed no active session
// if such cookie is not found in the request.
```

**Comment at lines 166-168:**
```go
// Notice that an "-aes" suffix is being used in the cookie names. This is for backwards compatibility and
// is/was meant to be able to differentiate between a session using cookies holding encrypted data, and the older
// less secure sessions using cookies holding JWTs.
```

### Evidence

Looking at the actual code at line 169:
```go
cookieName = sessionCookieName(SessionCookieName, s.Key)
```

Where `SessionCookieName = "kiali-token"` - **no "-aes" suffix is used anywhere**.

### Impact

- Misleading documentation
- Confusion for developers maintaining the code
- References to removed/deprecated functionality

### Suggested Fix

Update or remove the stale comments to accurately describe the current behavior.

---

### Problem

Different methods use different approaches to filter out nonce cookies:

| Function | Filter Used | Line |
|----------|-------------|------|
| `ReadAllSessions` | `!strings.HasPrefix(cookie.Name, NonceCookieName)` | 336 |
| `TerminateSession` | `!strings.Contains(cookie.Name, "nonce")` | 387 |

### Test Evidence

```
Filter comparison:
ReadAllSessions: !strings.HasPrefix(cookie.Name, "kiali-token-nonce")
TerminateSession: !strings.Contains(cookie.Name, "nonce")
```

**Behavior difference for hypothetical cookie `kiali-token-cluster-nonce-abc`:**
- `ReadAllSessions`: keeps it (`HasPrefix` doesn't match)
- `TerminateSession`: excludes it (`Contains` matches)

### Impact

- Inconsistent behavior between functions
- Potential edge cases where cookies are handled differently

### Suggested Fix

Standardize on `strings.Contains(cookie.Name, "nonce")` throughout for more robust filtering.

---

## Test Files Created

**File:** `kiali/handlers/authentication/session_persistor_bugs_test.go`

| Test Name | Purpose | Expected Result |
|-----------|---------|-----------------|
| `TestBug1_ReadAllSessions_DropsChunkCookies` | Proves chunk cookies are dropped | FAILS (bug exists) |
| `TestBug1_ReadAllSessions_FilterMatchesChunkCookies` | Shows filter logic bug | PASSES (correctly identifies bug) |
| `TestBug2_ChunksCookieNotKeyedInCreateSession` | Proves chunks cookie not keyed | FAILS (bug exists) |
| `TestBug2_TerminateSessionDoesNotFindKeyedChunksCookie` | Proves cleanup failure | FAILS (bug exists) |
| `TestBug2_MultipleKeyedSessionsCollideOnChunksCookie` | Proves session collision | PASSES (demonstrates collision) |
| `TestBug4_InconsistentNonceFiltering` | Shows inconsistent filtering | PASSES (documents inconsistency) |
| `TestHelper_sessionCookieName` | Verifies helper function | PASSES |

---

## Affected Code Locations

- `handlers/authentication/session_persistor.go`
  - `ReadAllSessions()` - lines 332-358
  - `CreateSession()` - lines 135-205 (specifically 188-202)
  - `readKialiCookie()` - lines 207-288 (specifically 224)
  - `TerminateSession()` - lines 380-391

---

## Additional Context

- These bugs become more impactful in **multi-cluster scenarios** where multiple keyed sessions exist
- **OpenShift auth controller** uses `ReadAllSessions` for session validation
- **Key rotation scenarios** may exacerbate the issue since cookies that fail to decrypt are dropped
- The existing test `TestTerminateSessionClearsAesSessionWithTwoChunks` at line 454 still uses "-aes" suffix in cookie names, suggesting the codebase may have been migrated but tests weren't updated

---

## Recommendations

1. **Priority 1 (Critical):** Fix Bug 1 - `ReadAllSessions` dropping chunk cookies causes immediate session corruption
2. **Priority 2 (High):** Fix Bug 2 - Chunks cookie collision affects multi-cluster deployments
3. **Priority 3 (Low):** Fix Bug 3 - Update stale comments
4. **Priority 4 (Low):** Fix Bug 4 - Standardize nonce filtering for consistency

### CODE REVIEW

**Comment by jmazzitelli** [comment]
notice the unit tests failed - which is expected. The first commit was unit tests that expose the bugs. There are no bug fixes in that first commit. I will commit another change that includes the fixes and the unit tests should go green.

**Comment by jmazzitelli** [comment]
since PR #7366 ("Implement multi-session for openshift auth strategy"), the `-aes` suffix was removed and the cookie name became just `kiali-token`. There are some comments hanging around that need to be cleaned up about this.

File: handlers/authentication/session_persistor.go [hunk 1]
```diff
@@ -329,21 +371,25 @@ func (p *cookieSessionPersistor[T]) ReadSession(r *http.Request, w http.Response
 
 // ReadAllSessions reads all session cookies from the request and returns the session data.
 // Returns an ErrNotFound if no session cookies are found.
+//
+// This function attempts to read all cookies that could potentially be session cookies.
+// Cookies that fail to decrypt (e.g., chunk cookies containing partial data) are skipped
+// without being dropped. This is important because:
+//  1. Dropping chunk cookies would corrupt the parent session
+//  2. Session keys can end with numbers (e.g., "cluster-1"), making it impossible to
+//     distinguish between a main session cookie and a chunk cookie by name alone
```
Comment (nrfox) [inline handlers/authentication/session_persistor.go]: This doesn't make sense. If you have a cookie chunk by definition you don't have a main session cookie.

File: handlers/authentication/session_persistor_bugs_test.go
Comment (nrfox) [inline handlers/authentication/session_persistor_bugs_test.go:1]: A separate massive test file is not needed. We would just need one or two test cases added in the style of the existing tests to `session_persistor_tests.go` covering the scenario where both:
1. You have cookie chunks
2. You have a session per cluster

**Review by nrfox — COMMENTED** [review]
There's two different use cases here that when combined do cause a bug.

1. Data we save in cookies is too large from some auth providers so we chunk the data into separate cookies.
2. OpenShift auth requires an auth session cookie _per cluster_ so we save a session cookie post fixed with the cluster name.

If you were to combine these two today i.e. your auth provider returns data that is too large too fit in one cookie AND you are using multi-cluster openshift auth then you would hit this bug but afaik you would currently not hit 1. using OpenShift auth because the payload we save is small enough to fit in one cookie. That could change in the future but for right now I don't think anyone would hit this in practice.

**Comment by jmazzitelli** [comment]
OK, so based on  @nrfox review, it sounds like this isn't a problem in practice today (due to OpenShift auth not needing the large data requiring cooking-chunking).

We can refine this later but it isn't required to be fixed right now.

I'll look into that main cookie thing - I thought the "main cookie" had data like "there are 4 chunks" and the remaining were the chunks (clusterA, clusterA-1, clusterA-2, clusterA-3, clusterA-4). We can revert that.

**Comment by jmazzitelli** [comment]
OK, the first cookie isn't the one that says how many chunks there are - there is one global cookie that has it. That's a problem if one session needs chunks and another doesn't (or two need chunks) because that chunk cookie is global - there is only one (this is true for master branch - this PR fixes this problem by making the chunk count cookie unique per session)

### Example 1 (single session, chunked) — `ReadAllSessions` corrupts it

User has one large session that gets chunked. Browser sends these cookies:

- `kiali-token` = first chunk
- `kiali-token-1` = second chunk
- `kiali-token-chunks` = `2`

`ReadAllSessions` loops over *all* cookies and its filter matches anything starting with `kiali-token`, so it tries to “read a session” from:

- `kiali-token` (works)
- `kiali-token-1` (this is **not** a full session; `readKialiCookie("kiali-token-1", ...)` fails)
- `kiali-token-chunks` (also not a session; fails)

When those reads fail, `ReadAllSessions` **drops the cookie it was trying to read** (calls `dropCookie` on `kiali-token-1` / `kiali-token-chunks`). Now the real session is corrupted because the chunk cookie `kiali-token-1` no longer exist.

### Example 2 (two sessions, one chunked) — global `kiali-token-chunks` breaks the other session

Suppose there are two sessions in the request:

- Session A (key `cluster1`) is small:
  - `kiali-token-cluster1` (single cookie)

- Session B (key `cluster2`) is large and chunked:
  - `kiali-token-cluster2`
  - `kiali-token-cluster2-1`
  - `kiali-token-chunks` = `2`  (**global**)

Now you call `ReadSession(..., key="cluster1")` on `master`.

Inside `readKialiCookie("kiali-token-cluster1", ...)` it checks for `kiali-token-chunks`. Since it exists (because of cluster2), it incorrectly assumes **cluster1 is chunked** and tries to read `kiali-token-cluster1-1`, which does not exist. cluster1 session fails to restore even though its cookie is fine.

File: handlers/authentication/session_persistor.go (hunk 1 above)
Comment (jmazzitelli) [reply handlers/authentication/session_persistor.go]: I changed the comments.. "main session" cookie just means the 0th cookie (the first chunk in a chunked session; or the full session cookie for unchunked session).

File: handlers/authentication/session_persistor_bugs_test.go
Comment (jmazzitelli) [reply handlers/authentication/session_persistor_bugs_test.go:1]: session_persistor_tests.go has 5 new tests. No separate test file for this issue is in the PR now.

**Comment by jmazzitelli** [comment]
I cleaned this up a bit. Tests should pass. Comments have been modified. We can wait before we consider taking this out of draft and merging. But this is definitely a potential issue, if not now then in the future.

**Comment by jmazzitelli** [comment]
rebased on master

**Comment by jmazzitelli** [comment]
> In `session_persistor_test` you should ensure that all calls to NewSessionData have `cluster` set (formerly named key). I see that it is unset (left to be "") in several places. That's not really true in the real code. Set a cluster and make sure the tests still pass.

done

**Comment by jshaughn** [comment]
@jmazzitelli , should we rebase this against master and move forward with it?

**Comment by jmazzitelli** [comment]
I think this is ready - we have tests so things should confirmed to be working
